R/ard_compare.R

```diff
@@ -0,0 +1,193 @@
+    }
+  } else {
+    key_columns_expr <- rlang::get_expr(key_columns_quo)
+    key_columns_env <- rlang::quo_get_env(key_columns_quo)
+    key_columns_value <- tryCatch(
+      rlang::eval_tidy(
+        key_columns_expr,
+        data = NULL,
+        env = key_columns_env
+      ),
+      error = identity
+    )
+
+    if (!inherits(key_columns_value, "error") && is.character(key_columns_value)) {
+      key_columns <- key_columns_value
+    } else {
+      key_columns <- NULL
+      process_selectors(x, key_columns = !!key_columns_quo, env = environment())
+    }
+
+    if (!is.character(key_columns)) {
+      cli::cli_abort(
+        "The {.arg key_columns} argument must be a character vector.",
+        call = get_cli_abort_call()
+      )
+    }
+
+    key_columns <- unique(key_columns)
+    key_origin <- "user"
+  }
+
+  key_columns <- unique(key_columns)
+
+  if (rlang::is_empty(key_columns)) {
+    cli::cli_abort(
+      "At least one column must be supplied in {.arg key_columns}.",
+      call = get_cli_abort_call()
+    )
+  }
+
+  missing_x <- setdiff(key_columns, names(x))
+  missing_y <- setdiff(key_columns, names(y))
+  if (!rlang::is_empty(missing_x) || !rlang::is_empty(missing_y)) {
+    cli::cli_abort(
+      c(
+        "!" = "The provided {.arg key_columns} must exist in both ARDs.",
+        if (!rlang::is_empty(missing_x)) {
+          "x" <- "Missing in {.arg x}: {.val {missing_x}}."
+        },
+        if (!rlang::is_empty(missing_y)) {
+          "x" <- "Missing in {.arg y}: {.val {missing_y}}."
+        }
+      ),
+      call = get_cli_abort_call()
+    )
+  }
+
+  .check_key_identify_rows(x, "x", key_columns, key_origin)
+  .check_key_identify_rows(y, "y", key_columns, key_origin)
+
+  fmt_column <- if ("fmt_fun" %in% names(x) || "fmt_fun" %in% names(y)) {
+    "fmt_fun"
+  } else if ("fmt_fn" %in% names(x) || "fmt_fn" %in% names(y)) {
+    "fmt_fn"
+  } else {
+    "fmt_fun"
+  }
+
+  comparison_targets <- list(
+    stat = "stat",
+    fmt_fn = fmt_column,
+    warning = "warning",
+    error = "error"
+  )
+
+  comparison_columns <- unique(unlist(comparison_targets, use.names = FALSE))
+
+  x_selected <-
+    dplyr::select(
+      x,
+      dplyr::all_of(key_columns),
+      dplyr::any_of(comparison_columns)
+    )
+  y_selected <-
+    dplyr::select(
+      y,
+      dplyr::all_of(key_columns),
+      dplyr::any_of(comparison_columns)
+    )
+
+  for (column in comparison_columns) {
+    x_selected <- .ensure_column(x_selected, column)
+    y_selected <- .ensure_column(y_selected, column)
+  }
+
+  # .check_rows_not_in_x_y(x_selected, y_selected, key_columns)
+
+  comparison <-
+    dplyr::full_join(
+      x_selected,
+      y_selected,
+      by = key_columns,
+      suffix = c(".x", ".y")
+    )
+  mismatch_list <- lapply(comparison_targets,
+    .build_mismatches,
+    comparison = comparison,
+    key_columns = key_columns
+  )
+
+  names(mismatch_list) <- names(comparison_targets)
+
+  mismatch_list
```

Comment:
Lastly, the function will return the `results` object, and add a class onto this list.

After we get this settled, we will write a print method for class to make it nice.